### PR TITLE
[programmable transactions] Added type declarations

### PR DESCRIPTION
**Pull request by tnowacki**
- Added type/struct declarations
- Added some basic checks/infrastructure

Nothing here is final. I'm just trying to break things up a bit so they will be easier to land and change in time. I'm not exactly sure how best to block the transaction kind outside of tests without a CFG check. If anyone has suggestions let me know :)

(built on #7791) 

EDIT: I also am sticking with the 2D approach as I think it will make things a bit easier when dealing with `entry` functions that take a vector of objects as inputs. I can write up a bit more about this if/when we want to discuss it.

### CODE REVIEW

File: crates/sui-types/src/messages.rs [hunk 1]
```diff
@@ -231,65 +231,298 @@ pub enum SingleTransactionKind {
...
+            return Err(SuiError::DuplicateObjectRefInput);
+        }
+        Ok(input_arg_objects
+            .into_iter()
+            .chain(commands.iter().flat_map(|command| command.input_objects()))
+            .collect())
+    }
+
+    fn validity_check(&self) -> SuiResult {
+        if !cfg!(test) {
```
Comment (tnowacki) [inline crates/sui-types/src/messages.rs:448]: I'm not sure of a better way of disabling the new transaction type until the feature is all ready?

File: crates/sui-types/src/messages.rs [hunk 2]
```diff
@@ -231,65 +231,298 @@ pub enum SingleTransactionKind {
...
+            return Err(SuiError::Unknown(
```
Comment (tnowacki) [inline crates/sui-types/src/messages.rs:449]: Does the error kind matter here for compatibility/deployment? Should I give `SuiError::InvalidDecoding` or something instead?

**Comment by tnowacki** [comment]
> authority_server

@bmwill, I imagine this would be in `handle_transaction`? 

What errors do we expect to emit for things that are "too new"? That is, something that would be in some new version of Sui, but is not yet supported?

File: crates/sui-types/src/messages.rs [hunk 3]
```diff
@@ -231,65 +231,298 @@ pub enum SingleTransactionKind {
...
+}
+
+/// A single command in a programmable transaction.
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Serialize, Deserialize)]
+pub enum Command {
+    /// A call to either an entry or a public Move function
+    MoveCall(Box<ProgrammableMoveCall>),
+    /// (Vec<forall T:key+store. T>, address)
+    /// It sends n-objects to the specified address. These objects must have store
+    /// (public transfer) and the previous owner must be an address.
```
Comment (amnn) [inline crates/sui-types/src/messages.rs:316]: Re: "previous owner must be an address" is this true?  What about if the object was created fresh as a result of a previous move call?

File: crates/sui-types/src/messages.rs [hunk 4]
```diff
@@ -231,65 +231,298 @@ pub enum SingleTransactionKind {
...
+    TransferObjects(Vec<Argument>, Argument),
+    /// (&mut Coin<T>, u64) -> Coin<T>
+    /// It splits of some amount into a new coin
```
Comment (amnn) [inline crates/sui-types/src/messages.rs:319]:
```suggestion
    /// It splits off some amount into a new coin
```

File: crates/sui-types/src/messages.rs [hunk 5]
```diff
@@ -231,65 +231,298 @@ pub enum SingleTransactionKind {
...
+    SplitCoin(Argument, Argument),
+    /// (&mut Coin<T>, Vec<Coin<T>>)
+    /// It merges n-coins into the first coin
+    MergeCoins(Argument, Vec<Argument>),
+    /// Publishes a Move package
+    Publish(Vec<Vec<u8>>),
```
Comment (amnn) [inline crates/sui-types/src/messages.rs:325]: 🎉

File: crates/sui-types/src/messages.rs [hunk 6]
```diff
@@ -231,65 +231,298 @@ pub enum SingleTransactionKind {
...
+    pub function: Identifier,
+    /// The type arguments to the function.
+    pub type_arguments: Vec<TypeTag>,
+    /// The arguments to the function.
+    pub arguments: Vec<Argument>,
+}
+
+impl ProgrammableMoveCall {
+    fn input_objects(&self) -> Vec<InputObjectKind> {
+        vec![InputObjectKind::MovePackage(self.package)]
```
Comment (amnn) [inline crates/sui-types/src/messages.rs:362]: Do we need to include packages from type arguments here?

File: crates/sui-types/src/messages.rs [hunk 7]
```diff
@@ -536,6 +792,10 @@ impl Display for SingleTransactionKind {
             Self::Genesis(_) => {
                 writeln!(writer, "Transaction Kind: Genesis")?;
             }
+            Self::ProgrammableTransaction(p) => {
+                writeln!(writer, "Transaction Kind: Programmable")?;
```
Comment (amnn) [inline crates/sui-types/src/messages.rs:796]: Nit: Most of the transactions format this with an extra space

```suggestion
                writeln!(writer, "Transaction Kind : Programmable")?;
```

I have a PR out that tweaks the output for `Genesis` and `ChangeEpoch` to match (among other things).  I'm not particularly attached to that formatting, but I'm a fan of consistency.

**Review by amnn — APPROVED** [review]
Left a question about extracting input objects from type parameters, but otherwise LGTM.

This PR is going to conflict with the CheckpointPrologue/Clock stack.  Are you feeling charitable enough to let that go first? ;) I'll owe you one.

**Comment by tnowacki** [comment]
> handle_transaction

We have the checks in the transaction input checker. This will prevent it from being signed unless `cfg(test)` passes. Is there a reason to have a check anywhere else?

I can add one in `authority_server` for example, but it feels a bit redundant. Thoughts?

**Comment by bmwill** [comment]
No need to add more if you feel confident with the checks you've added

**Comment by tnowacki** [comment]
> No need to add more if you feel confident with the checks you've added

I think what we have is fine, it's just not great. And I didn't really understand our planned versioning story before last week. I think the main thing to decide for this PR:
Is there a canonical error for things that are "too new" from a versioning perspective? Right now I just have `SuiError::Unknown`

**Pull request merged**